File: python/lib/list.py
################################################################################
# Deep Copy in time O(n), space O(1). Used the following link for reference:
# https://www.tutorialcup.com/linked-list/clone-linked-list-next-random-pointer.htm
################################################################################
from collections.abc import Sequence

class List(Sequence):
    def __init__(self, *links):
        self._head = None

        if len(links) == 0:
            return

        if len(links) == 1 and isinstance(links[0], Link):
            self._head = links[0]
            return

        for n in reversed(links):
            link = Link(n)
            link.next = self._head
            self._head = link

    def __getitem__(self, i):
        if not self.head:
            raise IndexError("list index out of range")

        link = self.head
        for x in range(i):
            if not link.next:
                raise IndexError
            link = link.next

        return link

        while link is not None:
            result += 1
            link = self.head.next


    def __len__(self, i):
        result = 0
        link = self.head
        while link is not None:
            result += 1
            link = self.head.next

    @property
    def head(self):
        return self._head

    def copy(self):
        self._copy_links()
        self._copy_refs()

        current = self.head
        copy = self.head.next
        while current is not None:
            current_copy = current.next
            current.next = current.next.next

            if current_copy.next:
                current_copy.next = current_copy.next.next

            current = current.next

        return List(copy)

    def _copy_links(self):
        current = self.head
        while current is not None:
            new = Link(current.value)
            new.next = current.next
            current.next = new
            current = new.next

    def _copy_refs(self):
        current = self.head
        while current is not None:
            if current.ref:
                current.next.ref = current.ref.next

            current = current.next.next

class Link:
    def __init__(self, value):
        self._value = value
        self._next = None
        self._ref = None

    @property
    def value(self):
        return self._value

    @property
    def next(self):
        return self._next

    @next.setter
    def next(self, link):
        self._next = link

    @property
    def ref(self):
        return self._ref

    @ref.setter
    def ref(self, link):
        self._ref = link

File: c/src/list.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "list.h"

int copy_list(List* from, List* to) {
  Link* current = from->head;
  while(current) {
    int result = add_link(to, current->value);
    if(result != 0) {
      return result;
    }
      
    current = current->next;
  }

  return 0;
}
//
//    def copy(self):
//        self._copy_links()
//        self._copy_refs()
//
//        current = self.head
//        copy = self.head.next
//        while current is not None:
//            current_copy = current.next
//            current.next = current.next.next
//
//            if current_copy.next:
//                current_copy.next = current_copy.next.next
//
//            current = current.next
//
//        return List(copy)
//
//    def _copy_links(self):
//        current = self.head
//        while current is not None:
//            new = Link(current.value)
//            new.next = current.next
//            current.next = new
//            current = new.next
//
//    def _copy_refs(self):
//        current = self.head
//        while current is not None:
//            if current.ref:
//                current.next.ref = current.ref.next
//
//            current = current.next.next

int add_link(List* list, const char* value) {
  Link* link = (Link*) malloc(sizeof(Link));
  if(link == NULL) {
    fprintf(stderr, "Unable to allocate memory for link.\n");
    return -1;
  }

  link->value = malloc(strlen(value) + 1);
  if(link->value == NULL) {
    fprintf(stderr, "Unable to allocate memory for vallue.\n");
    free(link);
    link == NULL;
    return -2;
  }

  strcpy(link->value, value);
  link->next = NULL;
  link->ref = NULL;

  if(!list->head) {
    list->head = link;

  } else {
    Link* end = list->head;
    while(end->next != NULL) {
      end = end->next;
    }
    end->next = link;
  }
  return 0;
}

void free_list(List* list) {
  Link* tmp;
  while (list->head != NULL) {
    tmp = list->head;
    list->head = list->head->next;
    free(tmp->value);
    tmp->next = NULL;
    free(tmp);
  }
}

void set_ref(Link* from, Link* to) {
  from->ref = to;
}
